src/forcing/ForcingsEngine.cpp

```diff
@@ -0,0 +1,239 @@
+            size
+        };
+
+        for (size_type di = 0; di < size; di++) {
+            // Set current variable value for each divide
+            var_cache_.at({{time_current_index_, di, vi}}) = var_span[di];
+        }
+    }
+
+    time_current_index_++;
+};
+
+ForcingsEngine::~ForcingsEngine()
+{
+    finalize();
+}
+
+ForcingsEngine& ForcingsEngine::instance(
+    const std::string& init,
+    std::size_t time_start,
+    std::size_t time_end
+)
+{
+    if (instances_[init] == nullptr) {
+        instances_[init] = std::make_unique<ForcingsEngine>(init, time_start, time_end);
+    }
+
+    return *instances_[init];
+}
+
+void ForcingsEngine::check_runtime_requirements()
+{
+    
+    // Check that the python module is installed.
+    {
+        auto interpreter_ = utils::ngenPy::InterpreterUtil::getInstance();
+        try {
+            auto mod = interpreter_->getModule("NextGen_Forcings_Engine");
+            auto cls = mod.attr("BMIForcingsEngine").cast<py::object>();
+        } catch(std::exception& e) {
+            throw std::runtime_error{
+                "Failed to initialize ForcingsEngine: ForcingsEngine python module is not installed or is not properly configured. (" + std::string{e.what()} + ")"
+            };
+        }
+    }
+
+    // Check that the WGRIB2 environment variable is defined
+    {
+        const auto* wgrib2_exec = std::getenv("WGRIB2");
+        if (wgrib2_exec == nullptr) {
+            throw std::runtime_error{"Failed to initialize ForcingsEngine: $WGRIB2 is not defined"};
+        }
+    }
+}
+
+boost::span<const std::string> ForcingsEngine::outputs() const noexcept
+{
+    return var_outputs_;
+}
+
+auto ForcingsEngine::time_begin() const noexcept
+  -> const clock_type::time_point&
+{
+    return time_start_;
+}
+
+auto ForcingsEngine::time_end() const noexcept
+  -> const clock_type::time_point&
+{
+    return time_end_;
+}
+
+auto ForcingsEngine::time_step() const noexcept
+  -> const clock_type::duration&
+{
+    return time_step_;
+}
+
+auto ForcingsEngine::time_index(time_t ctime) const noexcept
+  -> size_type
+{
+    const auto epoch = clock_type::from_time_t(ctime);
+
+    if (epoch < time_begin() || epoch > time_end()) {
+        return bad_index;
+    }
+
+    const auto offset = (epoch - time_begin()) / time_step();
+    return offset;
+}
+
+auto ForcingsEngine::divide_index(const std::string& divide_id) const noexcept
+  -> size_type
+{
+
+    const auto  id_sep   = divide_id.find('-');
+    const char* id_split = id_sep == std::string::npos
+                           ? divide_id.data()
+                           : &divide_id[id_sep + 1];
+    const int   id_int   = std::atoi(id_split);
+
+    const auto pos = std::lower_bound(var_divides_.begin(), var_divides_.end(), id_int);
+    if (pos == var_divides_.end() || *pos != id_int) {
+        return bad_index;
+    }
+
+    // implicit cast to unsigned
+    return std::distance(var_divides_.begin(), pos);
```

Comment:
I'm a bit concerned about assumptions made in this indexing algorithm, but I'm not sure if it's warranted. Do we have a guarantee of the numerical uniqueness of every divide label in the hydrofabric? If we really need that, it should probably be getting automatically checked somewhere at startup.

Reply:
We should, but I'd have to double check. I'll take a look at implementing a check for uniqueness, but most likely directly in the Forcings Engine's python code (I _think_ it makes more sense there)

Reply:
@mikejohnson51 Does the Hydrofabric make this promise that the numeric part of each divide ID is unique?

Reply:
Documenting discussion from the stand-up today, Justin is implementing a uniqueness check. That's the last outstanding concern I have with this PR.

Reply:
Added in [48e724c](https://github.com/NOAA-OWP/ngen/pull/720/commits/48e724c5bdbca5aeae89bc33b6e793ff37a55e22) -- the approach uses a temporary `std::unordered_map<int, int>` which increments the value on each divide_id, and checks for >1 values. If it finds one, a `std::runtime_error` is thrown -- let me know if there's any concerns with that